File: tool/validate_levels.py
import heapq
from typing import List, Tuple, Dict, Optional


class TileType:
    NORMAL = "normal"
    START = "start"
    END = "end"
    TELEPORT = "teleport"
    KEY = "key"
    LOCK = "lock"


class Tile:
    def __init__(self, row, col, type=TileType.NORMAL, teleport_id=None):
        self.row = row
        self.col = col
        self.type = type
        self.teleport_id = teleport_id


class Level:
    def __init__(self, id, rows, cols, optimal_steps, allowed_steps, special_tiles):
        self.id = id
        self.rows = rows
        self.cols = cols
        self.optimal_steps = optimal_steps
        self.allowed_steps = allowed_steps
        self.special_tiles = special_tiles

    def get_special(self, r, c):
        for t in self.special_tiles:
            if t.row == r and t.col == c:
                return t
        return None


def manhattan(r1, c1, r2, c2):
    return abs(r1 - r2) + abs(c1 - c2)


def solve_a_star(level: Level) -> Optional[int]:
    start_tile = next(t for t in level.special_tiles if t.type == TileType.START)
    end_tile = next(t for t in level.special_tiles if t.type == TileType.END)

    teleport_map: Dict[str, List[Tuple[int, int]]] = {}
    key_count = sum(1 for t in level.special_tiles if t.type == TileType.KEY)
    for t in level.special_tiles:
        if t.type == TileType.TELEPORT and t.teleport_id:
            teleport_map.setdefault(t.teleport_id, []).append((t.row, t.col))

    def heuristic(r, c):
        return manhattan(r, c, end_tile.row, end_tile.col)

    start_state = (heuristic(start_tile.row, start_tile.col), 0, start_tile.row, start_tile.col, 0)
    open_heap = [start_state]
    visited: Dict[str, int] = {f"{start_tile.row}_{start_tile.col}_0": 0}
    counter = 0

    dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    while open_heap:
        priority, steps, r, c, keys = heapq.heappop(open_heap)
        counter += 1
        if counter > 500000:
            return None

        if r == end_tile.row and c == end_tile.col:
            return steps

        for dr, dc in dirs:
            nr, nc = r + dr, c + dc
            if nr < 0 or nr >= level.rows or nc < 0 or nc >= level.cols:
                continue

            new_keys = keys
            special = level.get_special(nr, nc)
            if special and special.type == TileType.LOCK:
                if keys <= 0:
                    continue
                new_keys = keys - 1
            elif special and special.type == TileType.KEY:
                new_keys = keys + 1
                if new_keys > key_count:
                    new_keys = key_count

            new_steps = steps + 1
            state_key = f"{nr}_{nc}_{new_keys}"
            if state_key in visited and visited[state_key] <= new_steps:
                continue
            visited[state_key] = new_steps
            heapq.heappush(open_heap, (new_steps + heuristic(nr, nc), new_steps, nr, nc, new_keys))

            special2 = level.get_special(nr, nc)
            if special2 and special2.type == TileType.TELEPORT and special2.teleport_id:
                pair = teleport_map.get(special2.teleport_id, [])
                for pr, pc in pair:
                    if pr == nr and pc == nc:
                        continue
                    tp_key = f"{pr}_{pc}_{new_keys}"
                    if tp_key in visited and visited[tp_key] <= new_steps:
                        continue
                    visited[tp_key] = new_steps
                    heapq.heappush(open_heap, (new_steps + heuristic(pr, pc), new_steps, pr, pc, new_keys))

    return None


def build_levels() -> List[Level]:
    levels: List[Level] = []
    idx = 0

    for i in range(10):
        sr, sc, er, ec = 0, 0, 7, 7
        if i % 3 == 1:
            sr, sc, er, ec = 0, 7, 7, 0
        elif i % 3 == 2:
            sr, sc, er, ec = 3, 0, 3, 7
        buffer = max(5, 10 - i)
        levels.append(_make_level(idx, sr, sc, er, ec, buffer, [], "basic"))
        idx += 1

    for i in range(10):
        tiles = []
        if i % 2 == 0:
            tiles = [
                Tile(2, 3, TileType.KEY),
                Tile(5, 4, TileType.LOCK),
            ]
        else:
            tiles = [
                Tile(4, 2, TileType.KEY),
                Tile(3, 5, TileType.LOCK),
            ]
        buffer = max(4, 8 - (i // 3))
        levels.append(_make_level(idx, 0, 0, 7, 7, buffer, tiles, "key+lock"))
        idx += 1

    for i in range(10):
        tiles = []
        tp_id = f"tp_{i}"
        if i % 2 == 0:
            tiles = [
                Tile(1, 3, TileType.TELEPORT, tp_id),
                Tile(6, 4, TileType.TELEPORT, tp_id),
            ]
        else:
            tiles = [
                Tile(2, 2, TileType.TELEPORT, tp_id),
                Tile(5, 5, TileType.TELEPORT, tp_id),
            ]
        buffer = max(3, 7 - (i // 4))
        levels.append(_make_level(idx, 0, 0, 7, 7, buffer, tiles, "teleport"))
        idx += 1

    for i in range(10):
        tp_id = f"tp_comb_{i}"
        tiles = [
            Tile(2, 2, TileType.TELEPORT, tp_id),
            Tile(5, 5, TileType.TELEPORT, tp_id),
            Tile(3, 3, TileType.KEY),
            Tile(4, 4, TileType.LOCK),
        ]
        buffer = max(3, 6 - (i // 5))
        levels.append(_make_level(idx, 0, 0, 7, 7, buffer, tiles, "teleport+key+lock"))
        idx += 1

    for i in range(10):
        x_id = f"x_{i}"
        y_id = f"y_{i}"
        tiles = [
            Tile(1, 1, TileType.TELEPORT, x_id),
            Tile(6, 6, TileType.TELEPORT, x_id),
            Tile(1, 6, TileType.TELEPORT, y_id),
            Tile(6, 1, TileType.TELEPORT, y_id),
            Tile(3, 2, TileType.KEY),
            Tile(4, 5, TileType.LOCK),
            Tile(2, 5, TileType.KEY),
            Tile(5, 2, TileType.LOCK),
        ]
        buffer = max(2, 5 - (i // 5))
        levels.append(_make_level(idx, 0, 0, 7, 7, buffer, tiles, "double_teleport+2keys"))
        idx += 1

    return levels


def _make_level(idx, sr, sc, er, ec, buffer, extras, _):
    tiles = [
        Tile(sr, sc, TileType.START),
        Tile(er, ec, TileType.END),
        *extras,
    ]
    temp = Level(f"level_{idx + 1}", 8, 8, 0, 0, tiles)
    optimal = solve_a_star(temp)
    if optimal is None:
        optimal = 14
    return Level(
        f"level_{idx + 1}",
        8,
        8,
        optimal,
        optimal + buffer,
        tiles,
    )


def main():
    levels = build_levels()
    print(f"Total levels: {len(levels)}")
    failures = 0
    for i, lv in enumerate(levels):
        computed = solve_a_star(lv)
        match = computed == lv.optimal_steps
        if not match:
            failures += 1
        type_info = []
        for t in lv.special_tiles:
            if t.type != TileType.START and t.type != TileType.END:
                type_info.append(t.type)
        print(
            f"Level {i + 1:3d}: optimal={lv.optimal_steps:3d}, "
            f"allowed={lv.allowed_steps:3d}, computed={computed}, "
            f"match={match}, types=[{', '.join(type_info)}]"
        )

    print()
    print(f"Failures: {failures} / {len(levels)}")
    if failures == 0:
        print("ALL CHECKS PASSED")


if __name__ == "__main__":
    main()

File: tool/validate_levels.dart
// ignore_for_file: avoid_print
import 'package:collection/collection.dart';

enum TileType { normal, start, end, teleport, key, lock, bridge, flip }

class Tile {
  final int row;
  final int col;
  final TileType type;
  final String? teleportId;
  Tile({required this.row, required this.col, this.type = TileType.normal, this.teleportId});
}

class Level {
  final String id;
  final int rows;
  final int cols;
  final int optimalSteps;
  final int allowedSteps;
  final List<Tile> specialTiles;
  Level({
    required this.id,
    this.rows = 8,
    this.cols = 8,
    required this.optimalSteps,
    required this.allowedSteps,
    this.specialTiles = const [],
  });

  Tile? getSpecialTile(int r, int c) {
    for (final t in specialTiles) {
      if (t.row == r && t.col == c) return t;
    }
    return null;
  }
}

class _Node implements Comparable<_Node> {
  final int r;
  final int c;
  final int steps;
  final int keys;
  final int heuristic;

  _Node(this.r, this.c, this.steps, this.keys, this.heuristic);

  @override
  int compareTo(_Node other) =>
      (steps + heuristic) - (other.steps + other.heuristic);
}

int manhattan(int r1, int c1, int r2, int c2) =>
    (r1 - r2).abs() + (c1 - c2).abs();

int? solveAStar(Level level) {
  final startTile =
      level.specialTiles.firstWhere((t) => t.type == TileType.start);
  final endTile =
      level.specialTiles.firstWhere((t) => t.type == TileType.end);

  final teleportMap = <String, List<(int, int)>>{};
  final keyCount =
      level.specialTiles.where((t) => t.type == TileType.key).length;
  for (final t in level.specialTiles) {
    if (t.type == TileType.teleport && t.teleportId != null) {
      teleportMap.putIfAbsent(t.teleportId!, () => []).add((t.row, t.col));
    }
  }

  int heuristic(int r, int c) => manhattan(r, c, endTile.row, endTile.col);

  final open = PriorityQueue<_Node>();
  final visited = <String, int>{};

  final start =
      _Node(startTile.row, startTile.col, 0, 0, heuristic(startTile.row, startTile.col));
  open.add(start);
  visited['${start.r}_${start.c}_0'] = 0;

  final dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)];

  while (open.isNotEmpty) {
    final current = open.removeFirst();
    if (current.r == endTile.row && current.c == endTile.col) {
      return current.steps;
    }
    for (final (dr, dc) in dirs) {
      final nr = current.r + dr;
      final nc = current.c + dc;
      if (nr < 0 || nr >= level.rows || nc < 0 || nc >= level.cols) continue;

      int newKeys = current.keys;
      final special = level.getSpecialTile(nr, nc);
      if (special?.type == TileType.lock) {
        if (current.keys <= 0) continue;
        newKeys = current.keys - 1;
      } else if (special?.type == TileType.key) {
        newKeys = current.keys + 1;
        if (newKeys > keyCount) newKeys = keyCount;
      }

      final newSteps = current.steps + 1;
      final stateKey = '${nr}_${nc}_$newKeys';
      if (visited.containsKey(stateKey) && visited[stateKey]! <= newSteps) {
        continue;
      }
      visited[stateKey] = newSteps;
      open.add(_Node(nr, nc, newSteps, newKeys, heuristic(nr, nc)));

      final special2 = level.getSpecialTile(nr, nc);
      if (special2 != null &&
          special2.type == TileType.teleport &&
          special2.teleportId != null) {
        final pair = teleportMap[special2.teleportId!];
        if (pair != null) {
          for (final (pr, pc) in pair) {
            if (pr == nr && pc == nc) continue;
            final tpKey = '${pr}_${pc}_$newKeys';
            if (visited.containsKey(tpKey) &&
                visited[tpKey]! <= newSteps) continue;
            visited[tpKey] = newSteps;
            open.add(_Node(pr, pc, newSteps, newKeys, heuristic(pr, pc)));
          }
        }
      }
    }
  }
  return null;
}

List<Level> buildLevels() {
  final levels = <Level>[];
  int idx = 0;

  for (int i = 0; i < 10; i++) {
    int sr = 0, sc = 0, er = 7, ec = 7;
    if (i % 3 == 1) {
      sr = 0; sc = 7; er = 7; ec = 0;
    } else if (i % 3 == 2) {
      sr = 3; sc = 0; er = 3; ec = 7;
    }
    int buffer = 10 - i;
    if (buffer < 5) buffer = 5;
    levels.add(_makeLevel(idx++, sr, sc, er, ec, buffer, []));
  }

  for (int i = 0; i < 10; i++) {
    final tiles = <Tile>[];
    int kr, kc, lr, lc;
    if (i % 2 == 0) {
      kr = 2; kc = 3; lr = 5; lc = 4;
    } else {
      kr = 4; kc = 2; lr = 3; lc = 5;
    }
    tiles.add(Tile(row: kr, col: kc, type: TileType.key));
    tiles.add(Tile(row: lr, col: lc, type: TileType.lock));
    int buffer = 8 - (i ~/ 3);
    if (buffer < 4) buffer = 4;
    levels.add(_makeLevel(idx++, 0, 0, 7, 7, buffer, tiles));
  }

  for (int i = 0; i < 10; i++) {
    final tiles = <Tile>[];
    String tpId = 'tp_$i';
    if (i % 2 == 0) {
      tiles.add(Tile(row: 1, col: 3, type: TileType.teleport, teleportId: tpId));
      tiles.add(Tile(row: 6, col: 4, type: TileType.teleport, teleportId: tpId));
    } else {
      tiles.add(Tile(row: 2, col: 2, type: TileType.teleport, teleportId: tpId));
      tiles.add(Tile(row: 5, col: 5, type: TileType.teleport, teleportId: tpId));
    }
    int buffer = 7 - (i ~/ 4);
    if (buffer < 3) buffer = 3;
    levels.add(_makeLevel(idx++, 0, 0, 7, 7, buffer, tiles));
  }

  for (int i = 0; i < 10; i++) {
    final tiles = <Tile>[];
    String tpId = 'tp_comb_$i';
    tiles.add(Tile(row: 2, col: 2, type: TileType.teleport, teleportId: tpId));
    tiles.add(Tile(row: 5, col: 5, type: TileType.teleport, teleportId: tpId));
    tiles.add(Tile(row: 3, col: 3, type: TileType.key));
    tiles.add(Tile(row: 4, col: 4, type: TileType.lock));
    int buffer = 6 - (i ~/ 5);
    if (buffer < 3) buffer = 3;
    levels.add(_makeLevel(idx++, 0, 0, 7, 7, buffer, tiles));
  }

  for (int i = 0; i < 10; i++) {
    final tiles = <Tile>[];
    String x = 'x_$i';
    String y = 'y_$i';
    tiles.add(Tile(row: 1, col: 1, type: TileType.teleport, teleportId: x));
    tiles.add(Tile(row: 6, col: 6, type: TileType.teleport, teleportId: x));
    tiles.add(Tile(row: 1, col: 6, type: TileType.teleport, teleportId: y));
    tiles.add(Tile(row: 6, col: 1, type: TileType.teleport, teleportId: y));
    tiles.add(Tile(row: 3, col: 2, type: TileType.key));
    tiles.add(Tile(row: 4, col: 5, type: TileType.lock));
    tiles.add(Tile(row: 2, col: 5, type: TileType.key));
    tiles.add(Tile(row: 5, col: 2, type: TileType.lock));
    int buffer = 5 - (i ~/ 5);
    if (buffer < 2) buffer = 2;
    levels.add(_makeLevel(idx++, 0, 0, 7, 7, buffer, tiles));
  }

  return levels;
}

Level _makeLevel(int idx, int sr, int sc, int er, int ec, int buffer, List<Tile> extras) {
  final tiles = <Tile>[
    Tile(row: sr, col: sc, type: TileType.start),
    Tile(row: er, col: ec, type: TileType.end),
    ...extras,
  ];
  final temp = Level(
    id: 'level_${idx + 1}',
    rows: 8,
    cols: 8,
    optimalSteps: 0,
    allowedSteps: 0,
    specialTiles: tiles,
  );
  final optimal = solveAStar(temp) ?? 14;
  return Level(
    id: 'level_${idx + 1}',
    rows: 8,
    cols: 8,
    optimalSteps: optimal,
    allowedSteps: optimal + buffer,
    specialTiles: tiles,
  );
}

void main() {
  final levels = buildLevels();
  print('Total levels: ${levels.length}');
  int failures = 0;
  for (int i = 0; i < levels.length; i++) {
    final lv = levels[i];
    final computed = solveAStar(lv);
    final match = computed == lv.optimalSteps;
    if (!match) failures++;
    final typeInfo = <String>[];
    for (final t in lv.specialTiles) {
      if (t.type != TileType.start && t.type != TileType.end) {
        typeInfo.add(t.type.name);
      }
    }
    print('Level ${i + 1}: optimal=${lv.optimalSteps}, allowed=${lv.allowedSteps}, computeCheck=$computed, types=[${typeInfo.join(',')}]');
  }
  print('');
  print('Failures: $failures / ${levels.length}');
  if (failures == 0) print('ALL CHECKS PASSED');
}
